src/main/scala/fs2/kafka/TransactionalProducerMessage.scala

```diff
@@ -0,0 +1,208 @@
+    override val passthrough: P,
+    override val traverse: NonEmptyTraverse[G]
+  ) extends TransactionalProducerMessage[F, G, K, V, P] {
+    override def toString: String = {
+      implicit val G: Foldable[G] = traverse
+      records.mkString(
+        "TransactionalProducerMessage(",
+        ", ",
+        s", $offsetBatch, $consumerGroupId, $passthrough)"
+      )
+    }
+  }
+
+  /**
+    * Enables creating [[TransactionalProducerMessage]]s with the following syntax.
+    *
+    * {{{
+    * TransactionalMessage[G].of(records, passthrough)
+    *
+    * TransactionalMessage[G].of(records)
+    * }}}
+    */
+  final class ApplyBuilders[G[+ _]] private[kafka] (
+    private val G: NonEmptyTraverse[G]
+  ) extends AnyVal {
+
+    /**
+      * Creates a new [[TransactionalProducerMessage]] for producing zero or more
+      * `ProducerRecord`s within a Kafka transaction, then emitting a
+      * [[ProducerResult]] with the results and specified passthrough value.
+      */
+    def of[F[_], K, V, P](
+      records: G[(ProducerRecord[K, V], CommittableOffset[F])],
+      passthrough: P
+    )(implicit F: MonadError[F, Throwable]): F[TransactionalProducerMessage[F, G, K, V, P]] =
+      TransactionalProducerMessage(records, passthrough)(F, G)
+
+    /**
+      * Creates a new [[TransactionalProducerMessage]] for producing zero or more
+      * `ProducerRecord`s within a Kafka transaction, then emitting a
+      * [[ProducerResult]] with the results and `Unit` passthrough value.
+      */
+    def of[F[_], K, V](
+      records: G[(ProducerRecord[K, V], CommittableOffset[F])]
+    )(implicit F: MonadError[F, Throwable]): F[TransactionalProducerMessage[F, G, K, V, Unit]] =
+      TransactionalProducerMessage(records)(F, G)
+  }
+
+  /**
+    * Creates a new [[TransactionalProducerMessage]] for producing exactly one
+    * `ProducerRecord` within a Kafka transaction, then emitting a
+    * [[ProducerResult]] with the results and specified passthrough value.
+    */
+  def one[F[_], K, V, P](
+    record: (ProducerRecord[K, V], CommittableOffset[F]),
+    passthrough: P
+  )(implicit F: MonadError[F, Throwable]): F[TransactionalProducerMessage[F, Id, K, V, P]] =
+    apply[F, Id, K, V, P](record, passthrough)
+
+  /**
+    * Creates a new [[TransactionalProducerMessage]] for producing exactly one
+    * `ProducerRecord` within a Kafka transaction, then emitting a
+    * [[ProducerResult]] with the results and `Unit` passthrough value.
+    */
+  def one[F[_], K, V](
+    record: (ProducerRecord[K, V], CommittableOffset[F])
+  )(implicit F: MonadError[F, Throwable]): F[TransactionalProducerMessage[F, Id, K, V, Unit]] =
+    one(record, ())
+
+  /**
+    * Creates a new [[TransactionalProducerMessage]] for producing zero or more
+    * `ProducerRecord`s within a Kafka transaction, then emitting a
+    * [[ProducerResult]] with the results and specified passthrough value.
+    */
+  def apply[F[_], G[+ _], K, V, P](
```

Comment:
With this approach, we have to have exactly one `ProducerRecord` for a `CommittableOffset`. I think there is a case that you might actually have zero or more `ProducerRecord`s for every `CommittableOffset`, but currently we can't support that. Do you have any thoughts on this and how we could support it?

Reply:
Ergh, good point... I knew this would drop the zero-`ProducerRecord`s case (which seems OK since transactions are designed for consume-process-produce flows), but hadn't considered the many-`ProducerRecord`s case.

I could see two approaches to a different API:
1. Change this to take both `records: G[ProducerRecord[K, V]]` and `offsets: G[CommittableOffset[F]]`. This lets us avoid changing anything in `CommittableOffsetBatch`, but it forces the same `G` context for both args (which might not always make sense?)
2. Change this to take `records: G[ProducerRecord[K, V]]` and `offsetBatch: CommittableOffsetBatch[F]`, and add tracking of group IDs to `CommittableOffsetBatch` like you previously suggested. To get around the concerns I raised before I think the batch would need to track something like `consumerGroupIds: Set[String]` instead of just `consumerGroupId: Option[String]`. `TransactionalProducerMessage.apply` could enforce that the set contains exactly one element, but other usage sites could be OK with group ID sets of zero / many elements.

I lean towards trying option 2. WDYT?

Reply:
I gave this some thought, and if we look at it from a user point-of-view, you would use it something along the following lines. I'm assuming we want to commit multiple offsets per transaction, so you might use `Stream#chunks` or similar for chunking.

```scala
messages: Chunk[CommittableMessage[F, K, V]] =>
  TransactionalProducerMessage(
    records = messages.flatMap(message => Chunk(ProducerRecord(...))), // Technically we need NonEmptyTraverse
    offsetBatch = CommittableOffsetBatch.fromFoldable(messages.map(_.committableOffset)),
    passthrough = ???
  ).flatMap(producer.produce)
```

Two things which I think we could improve on here are the following.

1. We `map` over `messages` twice and there's a disconnect between which offsets are for which `ProducerRecord`s. It doesn't matter for the technical implementation, but for a user, I think it's nice to only have to do one `map` and see which offset is for which `ProducerRecord`s.
2. Due to the error handling being on `TransactionalProducerMessage`, we have to `flatMap`. This is not a big annoyance, but it results in less straightforward code than if error handling was done in `producer.produce`, since you wouldn't need to `flatMap` in that case.

If we wrote the equivalent code but taking those two points into account, we would end up with something like the following.

```scala
messages: Chunk[CommittableMessage[F, K, V]] =>
  producer.produce(
    TransactionalProducerMessage(
      messages.map { message =>
        CommittableProducerRecords(
          Chunk(ProducerRecord(...)),
          message.committableOffset
        )
      },
      passthrough = ???
    )
  )
```

The introduction of `CommittableProducerRecords` means we can see which offset is for which `ProducerRecord`s, and there's now only one `map`. Error handling is in `producer.produce` and I think we can handle the empty case with `F.unit` (since there's nothing to do).

`TransactionalProducerMessage` can do the offset batching using the exsiting `Traverse[G]` instance, but the slightly more tricky part is that we have `G[CommittableProducerRecords[F, H, K, V]]`, from which we can get `G[H[ProducerRecord[K, V]]`. This has to be 'flattened' to `G[ProducerRecord[K, V]]`. I think we can require `H = G`, so we can flatten easily. If type inference works, we can even have a `CommittableProducerRecords.one` which does the wrapping with `G` using `Applicative[G]`.

The added benefit is that some offsets in our transaction can now not result in producing some records, which might not be common, but is nice to support.

What do you think? I'm also happy to take a stab at this myself unless you want to.

Reply:
I think trying to reduce the number of `map` calls needed makes sense :+1: `CommittableProducerRecords` might also "fix" the issue of handling messages from multiple consumer groups in one transaction, if we have code within the `TransactionalKafkaProducer` like:
```scala
message: TransactionalProducerMessage =>
  message.committableRecords.traverse { records =>
    underlyingProducer.sendOffsetsToTransaction(
      records.offset.offsets,
      records.offset.consumerGroupId
    )
  }
```
... But that loses the efficiency of batching the offsets, so we might still want that as an internal pre-processing step.

> we can handle the empty case with F.unit (since there's nothing to do).

I think in this case we'd still need to call `commit` on the underlying offsets. Otherwise users will end up needing to branch the empty & non-empty cases anyways to ensure all the input records get committed.

I won't have much time the next few days to work on this PR, but I'm happy to try it out later if you don't get to it first 😄

Reply:
> ... But that loses the efficiency of batching the offsets, so we might still want that as an internal pre-processing step.

It's a good point. It should be rare to mix records from multiple consumers (`CommittableOffsetBatch` cannot even handle it properly at the moment). It's a good idea to guard against it though, and hopefully we can have `CommittableOffsetBatch` guard against it in the future as well.

> I think in this case we'd still need to call `commit` on the underlying offsets. Otherwise users will end up needing to branch the empty & non-empty cases anyways to ensure all the input records get committed.

We have two cases: (1) the `Chunk` is empty, so you end up with a `TransactionalProducerMessage` without any `CommittableProducerRecords` and only a passthrough, in which case you can do `F.pure(message.passthrough)` (this is the case I was referring to), (2) the `Chunk` is non-empty, but there are `CommittableProducerRecords` without records in them, in which case we have to `commit` (the case you're probably referring to).

> I won't have much time the next few days to work on this PR, but I'm happy to try it out later if you don't get to it first 😄

I'm quite interested to see if this works well, will try to find some time for it.

Reply:
I've pushed some WIP which passes the tests. Still need more tests, documentation, and general cleanup, but it gives an idea of the approach.

Reply:
I think this looks good, thanks for all the help! I started working on more tests, and I'm finding that the choice to require `G[_]` to be the same in both `CommittableRecords` and `TransactionalProducerMessage` can be a rough spot in usability. For example, the code I'd like to write is:
```scala
toProduce: List[(K, V, CommittableOffset[F])] =>
  val records = toProduce.map {
    case (key, value, offset) =>
      CommittableProducerRecords.one(ProducerRecord(topic, key, value), offset)
  }
  val message = TransactionalProducerMessage(records, "passthrough")
```
But `TransactionalProducerMessage.apply` doesn't compile because `G[_]` is set to `Id` from `CommittableProducerRecords.one`.

I'm not sure if it's worth adding another type parameter and dealing with the `G[H[ProducerRecord[K, V]]` problem, but I do think it makes `CommittableProducerRecords.one` much less useful. What do you think?

Reply:
It's a good point, and I thought about it while writing the implementation as well. I've pushed https://github.com/ovotech/fs2-kafka/pull/130/commits/88d17844ada00a6f9fdc3888fc29ccf37530b8ce which solves this by converting `H` to `G` using `Monad[G]`, `MonoidK[G]`, and `Foldable[H]`. It's not the most performant solution, but given you usually produce very few (~1) record per offset, I think the overhead isn't too bad here. It does prevent you from using 'non-empty' `G` collections though, due to the `MonoidK` requirement.

Reply:
I ended up changing `TransactionalProducerMessage` to only support `Chunk` for `G`, which solves the performance concerns. `Chunk` is going to be the the most used, if not only, type for `G` in practice, so think it's a reasonable restriction.